### PR TITLE
@ #492 | should mask specified sensitive values within log messages

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #492 
please check this PR, thank you!

### CODE REVIEW

**Comment by hieptranquoc** [comment]
config:
```
logger:
  masked_values: ['some value', 'another value']
```

File: lib/teracy-dev/logging/mask_log_manager.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module TeracyDev
+  module Logging
+    class MaskLogManager
+      @log_masked = {}
```
Comment (hoatle) [inline lib/teracy-dev/logging/mask_log_manager.rb]: masked_values

File: lib/teracy-dev/logging/mask_log_manager.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+      def self.add_log_masked(value)
```
Comment (hoatle) [inline lib/teracy-dev/logging/mask_log_manager.rb]: add_masked_values(values)

File: lib/teracy-dev/logging/mask_log_manager.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        if value.is_a? String
+          @log_masked[value] = '*' * value.length
+        else
+          value.each do |val|
+            @log_masked[val] = '*' * val.length
+          end
+        end
+      end
+
+      def self.mask_log(msg)
```
Comment (hoatle) [inline lib/teracy-dev/logging/mask_log_manager.rb]: def self.masked(msg)

File: lib/teracy-dev/loader.rb [hunk 4]
```diff
@@ -45,10 +46,14 @@ def sync
       entry_settings = @settingsManager.build_entry_settings(@extension_entry_path)
       settings = Util.override(teracy_dev_settings, entry_settings)
       @logger.debug("settings: #{settings}")
+      add_log_filter(settings['logger']['masked_values']) if Util.exist?(settings['logger']['masked_values'])
       sync_teracy_dev(settings['teracy-dev']['location'])
       sync_teracy_dev_entry(settings['teracy-dev']['entry_location'])
     end
 
+    def add_log_filter(value)
```
Comment (hoatle) [inline lib/teracy-dev/loader.rb]: def add_masked_values

**Review by hoatle — CHANGES_REQUESTED** [review]
@hieptranquoc pls check my comments, basically, this is what I'm expected, just need to name functions, variables more clearly and properly.

**Comment by hoatle** [comment]
Need to add `TeracyDev.add_logger_masked_values` api so that other extensions can call this api dynamically, don't forget to update this api docs: what it does and how to use, @since v0.6.0-a5 and any reference (for example, link to the related issue)

**Comment by hieptranquoc** [comment]
@hoale please check this PR again! thank you!

**Comment by hieptranquoc** [comment]
@hoatle please check this PR again! Thank you

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check this PR again. Thank you!

**Comment by hoatle** [comment]
did not work as expected:

- config_default.yaml

```yaml
variables:
  entry_secret: ${ENTRY_SECRET:-abcdef}
```

- logger.yaml

```yaml
logger:
  masked_values: ["%{entry_secret}"]
```

```shell
$ LOG_LEVEL=debug vagrant status | grep abcdef
[TeracyDevCore::Processors::Variables][DEBUG]: KEY: entry_secret, env_key: ENTRY_SECRET, env_value: , env_default: abcdef
```

expected: `env_default: ******`

```shell
$ ENTRY_SECRET=hello LOG_LEVEL=debug vagrant status | grep hello
[TeracyDevCore::Processors::Variables][DEBUG]: KEY: entry_secret, env_key: ENTRY_SECRET, env_value: hello, env_default: abcdef
```
expected: `env_value: *****`

- `@logger.debug("abcdef")` was not masked

```shell
$ LOG_LEVEL=debug vagrant status | grep abcdef
[TeracyDevCore::Processors::Variables][DEBUG]: KEY: entry_secret, env_key: ENTRY_SECRET, env_value: , env_default: abcdef
[TeracyDevCore::Config::Networks][DEBUG]: abcdefd
```

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check this PR again! Thank you!

**Comment by hieptranquoc** [comment]
original config:
```yaml
logger:
  mask:
    keys: {}
    values: []
```
example override config in config_override.yaml
```yaml
logger:
  mask:
    keys:
      node_name_prefix: ${NODE_NAME_PREFIX:-node}
      node_hostname_prefix: ${NODE_HOSTNAME_PREFIX:-k8s}
      node_domain_affix: ${NODE_DOMAIN_AFFIX:-local}
      iorad_app_github_username: ${IORAD_GITHUB_USERNAME:-tada-todo | tudu}
      iorad_app_github_password: "012345678"
      iorad_app_env_var: ${IORAD_APP_ENV_VAR}
    values: [settings, configure]
```

**Comment by hoatle** [comment]
@hieptranquoc still very buggy

logger.yaml:

```
logger:
  mask:
    variables:
      entry_secret: ${ENTRY_SECRET:-abcdef}
```

config_override.yaml

```
variables:
  entry_secret: ${ENTRY_SECRET:-hi}
```

with the following logger message for test:

`logger.debug("test hello abcdef hi")`

- without config_override.yaml:

expected: "test hello ****** hi" => OK

- with config_override.yaml and `$ LOG_LEVEL=debug vagrant status`
expected: "test hello abcdef **"
actual: "test hello ****** hi" => KO

- with config_override.yaml and `$ ENTRY_SECRET=hello LOG_LEVEL=debug vagrant status`
expected: "test *** abcdef hi"
actual: "test ***** ****** hi" => KO

**Comment by hoatle** [comment]
notes:
```
# - Build logger.mask
#   + load logger.yaml if any
#   + load config_override.yaml if any
#   + override "logger.mask.variables" from logger.yaml with "variables" from config_override.yaml if any
# - Resolve mask.variables to mask.values if logger.mask.variables exists then append values to MaskManager
#   + how to resolve: https://github.com/teracyhq-incubator/teracy-dev-core/blob/develop/lib/teracy-dev-core/processors/variables.rb#L14
# - Logger will ask MaskManager to mask its log message
```
notes:
```
- no teracy-dev-entry/logger.yaml => do nothing
- teracy-dev-entry/logger.yaml avaiable only => load and validate and then resolve logger.mask.variables
- teracy-dev-entry/config_override.yaml avaiable => load and validate and override logger.mask.variables from logger.yaml with variables from config_override.yaml and then resolve logger.mask.variables
```

**Comment by hoatle** [comment]
@hieptranquoc this case failed (the same setup from https://github.com/teracyhq/dev/pull/551#issuecomment-438885667)

with config_override.yaml and $ ENTRY_SECRET=hello LOG_LEVEL=debug vagrant status
expected: "test *** abcdef hi"
actual: "test ***** abcdef **" => KO

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check again. Thank you!

**Comment by hoatle** [comment]
@hieptranquoc this case failed when
config_override.yaml

```
variables:
  entry_secret: hi
```

expected: "test hello abcdef **"
actual: "test hello ****** **"

**Comment by hieptranquoc** [comment]
@hoatle I've updated! Please check again! Sr for wasting your time. Thank you!

**Review by hoatle — APPROVED** [review]
passed by basic test cases, will improve more later.

**Pull request merged**